README.md
# BT-0240 Configurator

Configurator for Bluetooth RS232 Adapter.
Device Original Manual with windows usage:
https://rainbow.com.ua/images/files/BT0240_User_Manual.pdf

Linux Usage:

bt0240 can print help:

    # ./bt0240 -h
    Usage: bt0240 [OPTION]...
    Configurator for BT-0240 Bluetooth RS232C Adapter

      -h, --help              Print help and exit
      -V, --version           Print version and exit
      -d, --device=STRING     Serial Device or MAC of BT-0240
                                (default=`/dev/ttyS0')
      -b, --bps=INT           Serial Speed in bps [9600-1382400]
                                (default=`115200')
          --parity=STRING     Serial Parity [none|even|odd]  (default=`none')
          --stopbits=INT      Serial Number of Stop Bits  (default=`1')
          --role=STRING       Bluetooth Role [master|slave]  (default=`slave')
          --discoverable=INT  Bluetooth Device Discoverable (0|1)  (default=`1')
          --peer=STRING       Bluetooth Peer MAC [AA:BB:CC:DD:EE:FF|any]
                                (default=`any')
      -p, --pin=STRING        Bluetooth Security PIN [4-12 chars|off]
                                (default=`1234')
          --name=STRING       Bluetooth Device Name [1-31 chars]  (default=`Serial
                                Adapter')
          --reset             Configure Device to Default Settings  (default=off)
      -v, --verbose           Print extra info
      -l, --list              List Device Settings

Connect 7.5V power supply and turn the switch ON.
Green LED &#x1f7e9; should light up.

Find out MAC address

    # hcitool scan
    Scanning ...
           00:02:72:11:22:33      Adapter1

Push "CFG" button near power plug socket.
Yellow LED &#x1f7e7; should light up, indicating
configuration is allowed now.
Run configurator to print current device settings:

    # ./bt0240 -d 00:02:72:11:22:33 -l
    Device name : Adapter1
    Encryption  : No
    Role        : Slave
    Discoverable: Yes
    Peer        : Any
    BPS Rate    : 9600
    Parity      : N
    Stopbits    : 1

Change Baudrate (Bits per second, BPS) to 19200:

    # ./bt0240 -d 00:02:72:11:22:33 --bps=19200

Print settings to see

    # ./bt0240 -d 00:02:72:11:22:33 -l
    Device name : Adapter1
    Encryption  : No
    Role        : Slave
    Discoverable: Yes
    Peer        : Any
    BPS Rate    : 19200
    Parity      : N
    Stopbits    : 1

To use device as RS232 serial port, power it OFF and ON,
Green LED &#x1f7e9; should be ON and Yellow LED &#x2b1b; should be OFF.
Then Serial port can connect on-demand to the slave:

      # rfcomm bind 14 00:02:72:11:22:33
      # ls /dev/rfcomm14
      /dev/rfcomm14

Use the port. When connected, Blue LED &#x1F7E6; should turn ON and blink on traffic.

      # screen /dev/rfcomm14 9600



changelog
ver 0.4.9:
	Add debian support


ver 0.4.8:
	cmdline parameters changes


bt0240.c
#include <stdio.h>   /* Standard input/output definitions */
#include <string.h>  /* String function definitions */
#include <unistd.h>  /* UNIX standard function definitions */
#include <fcntl.h>   /* File control definitions */
#include <errno.h>   /* Error number definitions */
#include <termios.h> /* POSIX terminal control definitions */
#include <malloc.h>  /* malloc() */
#include <sys/ioctl.h>
// #include <lockdev.h> /* serial device lock */
#include <sys/socket.h>
#include <bluetooth/bluetooth.h> /* address parser */
#include <bluetooth/rfcomm.h>

#include "bt0240.h"

int bt_rate2int[] = {9600, 9600, 19200, 38400, 57600, 115200, 230400, 460800, 921600, 1382400};
int bt_rate2int_n = sizeof(bt_rate2int) / sizeof(bt_rate2int[0]);
char bt_parity2char[] = {'N', 'O', 'E' };
int bt_stopbits2int[] = {1, 2};

int verbose = 0;

struct bt0240 bt0240_device;
struct bt0240 *bt0240 = &bt0240_device;

/* opens device, 
** sets serial parameters 
** and returns file descriptor id 
*/
int bt_open_serial(struct bt0240 *bt)
{
  int fd = -1;
  struct termios options;
  int retry = 0;
  char *device_name = bt->name;

  bt->fd = -1;
  for(retry = 0; fd < 0 && retry < 3; retry++)
  {
    fd = open(device_name, O_RDWR | O_NOCTTY | O_SYNC); /* O_NDELAY or O_NONBLOCK */
    if(fd >= 0)
      break;
    sleep(2);
  }
  bt->fd = fd;
  
  if (fd == -1) {
    perror("open_bt: Unable to open device");
    return -1;
  }
  
  fcntl(fd, F_SETFL, 0);
  
  /* Get the current options for the port */
  tcgetattr(fd, &options);

  /* Set the baud rates to 115200 */
  cfsetispeed(&options, B115200);
  cfsetospeed(&options, B115200);

  /* Enable the receiver and set local mode */
  options.c_cflag |= (CLOCAL | CREAD);

  options.c_cflag &= ~PARENB;
  options.c_cflag &= ~CSTOPB;
  options.c_cflag &= ~CSIZE;
  options.c_cflag |= CS8;

  options.c_iflag |= IGNBRK;
  options.c_iflag &= ~IGNBRK;
  options.c_iflag &= ~ICRNL;
  options.c_iflag &= ~INLCR;
  options.c_iflag &= ~IMAXBEL;
  options.c_iflag &= ~IXON;
  options.c_iflag &= ~IXOFF;

  options.c_oflag &= ~OPOST;
  options.c_oflag &= ~ONLCR;

  options.c_lflag &= ~ISIG;
  options.c_lflag &= ~ICANON;
  options.c_lflag &= ~IEXTEN;
  options.c_lflag &= ~ECHO;
  options.c_lflag &= ~ECHOE;
  options.c_lflag &= ~ECHOK;
  options.c_lflag |= NOFLSH;
  options.c_lflag &= ~ECHOCTL;
  options.c_lflag &= ~ECHOKE;

  options.c_cc[VTIME] = 0;
  options.c_cc[VMIN] = 1;
  /* Set the new options for the port */
  tcsetattr(fd, TCSANOW, &options);

  usleep(100000); /* port open grace time */

  return fd;
}

int bt_open_bluetooth(struct bt0240 *bt)
{
  struct sockaddr_rc addr = { 0 };
  int status;

  /* allocate a socket */
  bt->fd = socket(AF_BLUETOOTH, SOCK_STREAM, BTPROTO_RFCOMM);
  bt->connect = 0;

  if(bt->fd < 0)
  {
    perror("socket");
    return -1;
  }
  /* set the connection parameters (who to connect to) */
  addr.rc_family = AF_BLUETOOTH;
  addr.rc_channel = (uint8_t) 1;
  str2ba(bt->name, &addr.rc_bdaddr); 

  /* connect to server (bluetooth slave) */
  status = connect(bt->fd, (struct sockaddr *)&addr, sizeof(addr));
   
  if(status < 0)
  {
    /* perror("connect"); */
    return -1;
  }

  bt->connect = 1;
  return bt->fd;
}


int bt_close_serial(struct bt0240 *bt)
{
  if(bt->fd >= 0)
    close(bt->fd);

  return 0;
}

int bt_close_bluetooth(struct bt0240 *bt)
{
  if(bt->fd >= 0)
  {
    if(bt->connect)
      shutdown(bt->fd, SHUT_RDWR);
    close(bt->fd);
  }
  return 0;
}

int bt_close(struct bt0240 *bt)
{
  if(bt->name == NULL)
    return -1;

  if(bt->name[0] == '/')
    return bt_close_serial(bt);

  return bt_close_bluetooth(bt);  
}

int bt_open(struct bt0240 *bt)
{
  int i, retval;
  if(bt->name == NULL)
    return -1;

  if(bt->name[0] == '/')
    return bt_open_serial(bt);

  for(i = 0; i < 3; i++)
  {
    retval = bt_open_bluetooth(bt);
    if(retval >= 0)
      return retval;
    bt_close_bluetooth(bt);
    sleep(2);
  }
  return -1;
}

/* return value: */
/* <0 if error */
/* >=0 sucess, length of entire packet */
int bt_read_raw(struct bt0240 *bt, char *packet, int maxlen)
{
  char *command, *arg_len, *arg;
  int i, n, skip; /* i-retries, n - byte pointer */
  fd_set rfds;
  struct timeval tv;
  int retval;
  char expect;
  int fd = bt->fd;

  if(packet == NULL || maxlen == 0)
    return 0;
  
  expect = packet[0];

  /* first we want to read initially expected byte */
  for(i = 0, n = 0; i < 10 && n < 2; i++)
  {
    /* Watch fd to see when it has input. */
    FD_ZERO(&rfds);
    FD_SET(fd, &rfds);

    /* Wait up to 0.5 seconds. */
    tv.tv_sec = 1;
    tv.tv_usec = 300000;
  
    retval = select(fd+1, &rfds, NULL, NULL, &tv);
    /* Don't rely on the value of tv now! */
  
    if (retval == -1)
    {
      if(verbose)
        printf("select error\n");
      return -1;
    }
    else if (retval)
    {
      // printf("Header byte #%d is available now.\n", n);
    }
    /* FD_ISSET(fd, &rfds) will be true. */
    else
    {
      if(verbose)
        printf("timeout in header\n");
      return -1; // printf("No data within five seconds.\n");
    }    
    // fcntl(fd, F_SETFL, FNDELAY); /* don't block serial read */
    
    retval = read(fd, &(packet[n]), 1);
    // printf("char %d=%02x\n", n, packet[n]);
    if(retval > 0 && packet[0] == expect)
    {
      n++;
      if(n == 2 && packet[1] + 2 != maxlen) 
        n = 0; /* length doesn't match expected maxlen */
    }
  }
  skip = i;

  command = packet;
  arg_len = packet+1;
  arg     = packet+2;

  if(n < 0)
    return -1;

  if(n == 2 && *arg_len > 0 && *arg_len + 2 <= maxlen)
  {
    {
      /* Watch fd to see when it has input. */
      FD_ZERO(&rfds);
      FD_SET(fd, &rfds);

      /* Wait up to 0.2 seconds. */
      tv.tv_sec = 0;
      tv.tv_usec = 200000;
  
      retval = select(fd+1, &rfds, NULL, NULL, &tv);
      /* Don't rely on the value of tv now! */
    
      if (retval == -1)
      {
        if(verbose)
          printf("select error\n");
        return -1; // perror("select()");
      }
      else if (retval)
      {
        // printf("Data is available now.\n");
      }
      /* FD_ISSET(fd, &rfds) will be true. */
      else
      {
        if(verbose)
          printf("timeout in payload\n");
        return -1; // printf("No data within five seconds.\n");
      }
      usleep(96*maxlen); /* calculate 96us character time (70ms should be enoug) */
      n += read(fd, arg, *arg_len);
    }
  }
  
  if(verbose)
  {
    printf("skipped %d, ", skip-2);
    for(i = 0; i < n; i++)
      printf("%02x ", packet[i]);
    printf("n=%d bytes\n", n);
  }
  return n;
}

int bt_write_raw(struct bt0240 *bt, char *packet, int maxlen)
{
  int fd = bt->fd;
  fcntl(fd, F_SETFL, 0); /* block serial write */
  return write(fd, packet, maxlen);
}

int bt_copystring(char *packet, char *string, int maxlen)
{
  int len = 0;
  if((int)packet[0] > 0)
    len = (int)packet[0] < maxlen ? (int)packet[0] : maxlen;
  memcpy(string, packet + 1, len);
  string[len] = '\0';
  return len;
}

int flushstream(int fd)
{
  fd_set rfds;
  struct timeval tv;
  int retval, bytes;
  char packet[100];

  for(bytes = 1; bytes > 0;)
  {
    /* Watch fd to see when it has input. */
    FD_ZERO(&rfds);
    FD_SET(fd, &rfds);

    /* Wait up to 0.3 seconds for quiet line. */
    tv.tv_sec = 0;
    tv.tv_usec = 300000;
    
    retval = select(fd+1, &rfds, NULL, NULL, &tv);

    bytes = 0;
    if(retval)
      bytes = read(fd, packet, 1);
  }
  return 0;
}

int bt_send_receive(struct bt0240 *bt, char *query, int querylen, char *answer, int answerlen, int verify)
{
  int retry = 0;
  int read = 0;
  int result = 0;
  char *answer2 = NULL;
  int equal = 0, require_equal = 0;
  int fd = bt->fd;

  if(verbose)
    printf("send_receive\n");
  if(verify)
    answer2 = malloc(answerlen);

  if(answer2 != NULL && answerlen > 0)
  {
    if(verbose)
      printf("verify reads\n");
    memcpy(answer2, answer, answerlen);
    /* spoil beginning and the end */
    answer2[0]++;
    answer2[answerlen - 1]--;
    require_equal = 1; /* requre 2 consecutive answers equal */
  }

  for(retry = 0; retry < 10; retry++)
  {
    if(verbose)
      if(retry >= 0)
        printf("RETRY %d EQUAL=%d\n", retry, equal);
    if(retry == 5)
    {
      bt_close(bt);
      if(verbose)
        printf("Reopening port to revive the line\n");
      sleep(5);
      fd = bt_open(bt);
    }

#if 0
    tcflush(fd, TCIOFLUSH);
#endif
    flushstream(fd);
    if(bt_write_raw(bt, query, querylen) < 0)
    {
      bt_close(bt);
      if(verbose)
        printf("Reopening port to revive the line\n");
      sleep(1);
      fd = bt_open(bt);
      // goto bailout;
    }
    read = bt_read_raw(bt, answer, answerlen);
    equal++;
    if(read < answerlen)
      equal = 0;
    if(answer2 != NULL)
    {
      if(memcmp(answer, answer2, answerlen) != 0)
        equal = 0;
      memcpy(answer2, answer, answerlen);
    }
    if(equal >= require_equal)
      goto complete;
  }
  if(equal < require_equal)
    goto bailout;
  /* printf("\n"); */

 complete:;
  result = 1;

 bailout:;
  if(answer2 != NULL)
    free(answer2);
  return result;
}

int bt_command_receive(struct bt0240 *bt, char command, char *answer, int answerlen)
{
  struct bt_command btcommand;

  btcommand.command = command;
  btcommand.len = 0;
  answer[0] = command;
  return bt_send_receive(bt, (char *) &btcommand, sizeof(btcommand), answer, answerlen, 1);
}

int bt_read_security(struct bt0240 *bt, struct bt_security1 *security)
{
  return bt_command_receive(bt, READ_SECURITY, (char *)security, sizeof(*security) );
}

int bt_read_identity(struct bt0240 *bt, struct bt_identity *name)
{
  return bt_command_receive(bt, READ_IDENTITY, (char *)name, sizeof(*name) );
}

int bt_read_uart(struct bt0240 *bt, struct bt_uart *uart)
{
  return bt_command_receive(bt, READ_UART, (char *)uart, sizeof(*uart) );
}

int bt_read_role(struct bt0240 *bt, struct bt_role *role)
{
  return bt_command_receive(bt, READ_ROLE, (char *)role, sizeof(*role) );
}

int bt_read_change_verify(struct bt0240 *bt, char *send, int len_send, char *receive, int len_receive)
{
  int result;
  /* TODO:
  ** read current state (read if needed, should have 'dirty' flag),
  ** compare read data with the requested
  ** modify only if the change is needed.
  ** return status if the operation succeeded (set dirty flag to 0)
  */
  result = bt_send_receive(bt, send, len_send, receive, len_receive, 0);
  return 0;
}

int bt_set_identity(struct bt0240 *bt, char *name)
{
  struct bt_identity identity;
  struct bt_commit commit;
  int len;
  len = (char) strlen(name);
  if(len < 0)
    len = 0;
  if(len > LEN_IDENTITY_NAME)
    len = LEN_IDENTITY_NAME;
  identity.command = WRITE_IDENTITY;
  identity.len = (char) LEN_IDENTITY_NAME;
  memset(identity.name, 0, LEN_IDENTITY_NAME);
  memcpy(identity.name, name, len);

  commit.command = (char) WRITE_IDENTITY_COMMIT;

#if 0
  printf("Warning: after setting device name, power cycle the BT-0240\n");
#endif
  return bt_read_change_verify(bt, (char *) &identity, sizeof(identity), (char *) &commit, sizeof(commit));
}

int bt_set_uart(struct bt0240 *bt, int rate, char parity, int stopbits)
{
  struct bt_uart uart;
  struct bt_commit commit;
  int i;
  int uart_rate = 5, uart_parity = 0, uart_stopbits = 0;

  for(i = 1; i < bt_rate2int_n; i++)
  {
    if(bt_rate2int[i] == rate)
      uart_rate = i;
  }

  switch(parity)
  {
  case 'N':
  case 'n':
    uart_parity = 0;
    break;
  case 'O':
  case 'o':
    uart_parity = 1;
    break;
  case 'E':
  case 'e':
    uart_parity = 2;
    break;
  }
  if(stopbits >= 1 && stopbits <= 2)
    uart_stopbits = stopbits-1;

  uart.command = WRITE_UART;
  uart.len = 3;
  uart.rate = uart_rate;
  uart.parity = uart_parity;
  uart.stopbits = uart_stopbits;

  commit.command = (char) WRITE_UART_COMMIT;

  return bt_read_change_verify(bt, (char *) &uart, sizeof(uart), (char *) &commit, sizeof(commit));
}

int bt_set_security(struct bt0240 *bt, char *pin)
{
  struct bt_security1 security;
  struct bt_commit commit;
  int len;
  
  security.command = (char) WRITE_SECURITY;
  security.len = sizeof(security) - 2;
  security.encryption = pin != NULL ? 1 : 0;

  security.pin_len = (char) 0;
  memset(security.pin, 0, LEN_SECURITY_PIN);
  if(pin != NULL)
  {
    len = (char) strlen(pin);
    if(len < 0)
      len = 0;
    if(len > LEN_SECURITY_PIN)
      len = LEN_SECURITY_PIN;
    security.pin_len = (char) len;
    memcpy(security.pin, pin, len);
  }
  commit.command = (char) WRITE_SECURITY_COMMIT;  

  return bt_read_change_verify(bt, (char *) &security, sizeof(security), (char *) &commit, sizeof(commit));
}

int bt_set_role(struct bt0240 *bt, int master, int discoverable, char *peer)
{
  struct bt_role role;
  struct bt_commit commit;

  memset(role.peer, 0, LEN_ROLE_PEER);
  if(peer != NULL)
  {
    if(strlen(peer) == LEN_ROLE_PEER)
      memcpy(role.peer, peer, LEN_ROLE_PEER);
    else
      peer = NULL;
  }

  role.role = 
    (master > 0 ? ROLE_MASTER : 0) | 
    (discoverable > 0 ? ROLE_DISCOVERABLE : 0) | 
    (peer != NULL ? ROLE_CONTACT_PEER : 0);

  role.command = (char) WRITE_ROLE;
  role.len = sizeof(role) - 2;
  commit.command = (char) WRITE_ROLE_COMMIT;
  
  return bt_read_change_verify(bt, (char *) &role, sizeof(role), (char *) &commit, sizeof(commit));
}


bt0240.h
// #include <lockdev.h>
/* struct for exchanging messages with bt0240 adapter */

/* commands that can be sent to bt0240 */
#define WRITE_ROLE 1
#define READ_ROLE 2
#define LEN_ROLE_PEER 12
#define ROLE_MASTER 1
#define ROLE_DISCOVERABLE 2
#define ROLE_CONTACT_PEER 4
#define WRITE_ROLE_COMMIT 3

#define WRITE_SECURITY 4
#define READ_SECURITY 5
#define LEN_SECURITY_PIN 12
#define WRITE_SECURITY_COMMIT 6

#define WRITE_IDENTITY 7
#define READ_IDENTITY 8
#define LEN_IDENTITY_NAME 31
#define LEN_IDENTITY_NAME2 31
#define WRITE_IDENTITY_COMMIT 9

#define WRITE_UART 10
#define READ_UART 11
#define UART_4800 0
#define UART_9600 1
#define UART_19200 2
#define UART_38400 3
#define UART_57600 4
#define UART_115200 5
#define UART_230400 6
#define UART_460800 7
#define UART_921600 8
#define UART_1382400 9
#define UART_NONE 0
#define UART_ODD 1
#define UART_EVEN 2
#define UART_1STOP 0
#define UART_2STOP 1
#define WRITE_UART_COMMIT 12

struct bt_commit {
  char command;
  char len;
  char data[1];
};

struct bt_command {
  char command;
  char len;
};

struct bt_role {
  char command;
  char len;
  char role;     /* see ROLE_* flags */ 
  char peer[LEN_ROLE_PEER];
};

struct bt_uart {
  char command;
  char len;
  char rate;
  char stopbits;
  char parity;
};

struct bt_identity {
  char command;
  char len;
  char name[LEN_IDENTITY_NAME];
};

struct bt_identity2 {
  char command;
  char len;
  char name[LEN_IDENTITY_NAME2];
};

struct bt_security1 {
  char command;
  char len;
  char encryption;
  char pin_len;
  char pin[LEN_SECURITY_PIN];
};

struct bt0240 {
  int fd; /* open device file */
  char *name; /* device name */
  int connect;
};

extern int bt_rate2int[], bt_rate2int_n;
extern char bt_parity2char[];
extern int bt_stopbits2int[];
#if 0
extern int fd;
extern char *device;
#endif
extern struct bt0240 *bt0240;
extern int verbose;

int bt_open(struct bt0240 *bt);
int bt_close(struct bt0240 *bt);

int bt_copystring(char *packet, char *string, int maxlen);

int bt_read_security(struct bt0240 *bt, struct bt_security1 *security);
int bt_read_identity(struct bt0240 *bt, struct bt_identity *name);
int bt_read_uart(struct bt0240 *bt, struct bt_uart *uart);
int bt_read_role(struct bt0240 *bt, struct bt_role *role);

int bt_set_identity(struct bt0240 *bt, char *name);
int bt_set_uart(struct bt0240 *bt, int rate, char parity, int stopbits);
int bt_set_security(struct bt0240 *bt, char *pin);
int bt_set_role(struct bt0240 *bt, int master, int discoverable, char *peer);


main.c
/* main.c
**
** Configures BT-0240 Bluetooth RS232C Adapter
**
** License: GPL
**
** Emard
**
*/
#include <stdio.h>   /* Standard input/output definitions */
#include <string.h>  /* String function definitions */
#include <unistd.h>  /* UNIX standard function definitions */
#include <fcntl.h>   /* File control definitions */
#include <errno.h>   /* Error number definitions */
#include <termios.h> /* POSIX terminal control definitions */
// #include <sys/socket.h>
#include <bluetooth/bluetooth.h> /* address parser */
#include <bluetooth/rfcomm.h>

#include "bt0240.h"
#include "cmdline.h"

struct gengetopt_args_info args_info;
struct gengetopt_args_info *args = &args_info;

int fd;
char *device = "/dev/ttyS0";

int main(int argc, char **argv) 
{
  int i;
  struct bt_uart btuart;
  struct bt_role btrole;
  struct bt_identity btname;
  struct bt_security1 btsecurity;
  char string[256];
  char *peer = NULL;
  char *pin = NULL;
  char *name;
  int bps;
  char parity;
  int stopbits;
  int master = 0;
  int discoverable = 1;
  int set_role = 0, set_security = 0, set_uart = 0, set_name = 0;
  struct sockaddr_rc mac = { 0 };
  char mac_str[100];

  cmdline_parser(argc, argv, args);
  verbose = args->verbose_given ? 1 : 0;
  bt0240->name = args->device_arg;

  #if 0
  if(bt0240->name[0] == '/')
  {
    if(dev_lock(bt0240->name))
    {
      perror("can't get hold of device lock");
      return 1;
    }
  }
  #endif

  fd = bt_open(bt0240);

  if(fd < 0)
  {
    perror("can't open device");
    return 1;
  }

  if(args->peer_given || args->reset_given)
  {
    if(strcasecmp("any", args->peer_arg) == 0)
    {
      peer = NULL;
      set_role = 1;
    }
    if(strlen(args->peer_arg) == 12)
    {
      peer = args->peer_arg;
      set_role = 1;
    }
    if(strlen(args->peer_arg) == 17)
    {
      str2ba(args->peer_arg, &mac.rc_bdaddr);
      sprintf(mac_str, "%02x%02x%02x%02x%02x%02x", 
              mac.rc_bdaddr.b[5], mac.rc_bdaddr.b[4], mac.rc_bdaddr.b[3],
              mac.rc_bdaddr.b[2], mac.rc_bdaddr.b[1], mac.rc_bdaddr.b[0]);
      peer = mac_str;
      set_role = 1;
    }
  }
  
  if(args->discoverable_given || args->reset_given)
  {
    discoverable = args->discoverable_arg > 0 ? 1 : 0;
    set_role = 1;
  }

  if(args->role_given || args->reset_given)
  {
    if(strcasecmp("master", args->role_arg) == 0)
    {
      master = 1;
      set_role = 1;
    }
    if(strcasecmp("slave", args->role_arg) == 0)
    {
      master = 0;
      set_role = 1;
    }
  }

  if(args->pin_given || args->reset_given)
  {
    if(strcasecmp("off", args->pin_arg) == 0)
    {
      pin = NULL;
      set_security = 1;
    }
    if(strlen(args->pin_arg) >= 4 && strlen(args->pin_arg) <= 12)
    {
      pin = args->pin_arg;
      set_security = 1;
    }
  }
  
  if(args->bps_given || args->reset_given)
  {
    int i;

    for(i = 0; i < bt_rate2int_n; i++)
    {
      if(args->bps_arg == bt_rate2int[i])
      {
        bps = args->bps_arg;
        set_uart = 1;
      }
    }
  }

  if(args->parity_given || args->reset_given)
  {
    if(strcasecmp("none", args->parity_arg) == 0 ||
       strcasecmp("n", args->parity_arg) == 0)
    {
      parity = 'N';
      set_uart = 1;
    }
    if(strcasecmp("even", args->parity_arg) == 0 ||
       strcasecmp("e", args->parity_arg) == 0)
    {
      parity = 'E';
      set_uart = 1;
    }
    if(strcasecmp("odd", args->parity_arg) == 0 ||
       strcasecmp("o", args->parity_arg) == 0)
    {
      parity = 'O';
      set_uart = 1;
    }
  }

  if(args->stopbits_given || args->reset_given)
  {
    if(args->stopbits_arg >= 1 && args->stopbits_arg <= 2)
    {
      stopbits = args->stopbits_arg;
      set_uart = 1;
    }
  }

  if(args->name_given || args->reset_given)
  {
    if(strlen(args->name_arg) >= 1 && strlen(args->name_arg) <= 31)
    {
      name = args->name_arg;
      set_name = 1;
    }
  }

  if(set_role)
    bt_set_role(bt0240, master, discoverable, peer);

  if(set_security)
    bt_set_security(bt0240, pin);

  if(set_uart)
    bt_set_uart(bt0240, bps, parity, stopbits);

  if(set_name)
  {
    printf("Warning: before changing the name next time, power cycle BT-0240\n");
    bt_set_identity(bt0240, name);
  }

  if(args->list_given)
  for(i = 0; i < 1; i++)
  {
    if(bt_read_identity(bt0240, &btname))
    {
      bt_copystring((char *) &(btname.len), string, sizeof(btname.name));
      printf("%-12s: %s\n", "Device name", string);
    }
    else
      printf("%-12s: *READ ERROR*\n", "Device name");

    if(bt_read_security(bt0240, &btsecurity))
    {
      bt_copystring((char *) &(btsecurity.pin_len), string, sizeof(btsecurity.pin));
      printf("%-12s: %s\n", "Encryption", (int)btsecurity.encryption > 0 ? "Yes" : "No");
      if((int)btsecurity.encryption > 0)
        printf("%-12s: %s\n", "PIN", string);
    }
    else
      printf("%-12s: *READ ERROR*\n%-12s: *READ ERROR*\n", "Encryption", "PIN");

    if(bt_read_role(bt0240, &btrole))
    {
      strcpy(string, "xx:xx:xx:xx:xx:xx");
      if( (btrole.role & ROLE_CONTACT_PEER) > 0)
      {
        string[0]  = btrole.peer[0];
        string[1]  = btrole.peer[1];

        string[3]  = btrole.peer[2];
        string[4]  = btrole.peer[3];

        string[6]  = btrole.peer[4];
        string[7]  = btrole.peer[5];

        string[9]  = btrole.peer[6];
        string[10] = btrole.peer[7];

        string[12] = btrole.peer[8];
        string[13] = btrole.peer[9];

        string[15] = btrole.peer[10];
        string[16] = btrole.peer[11];
      }    
      printf("%-12s: %s\n%-12s: %s\n%-12s: %s\n",
             "Role", btrole.role & ROLE_MASTER ? "Master" : "Slave",
             "Discoverable", btrole.role & ROLE_DISCOVERABLE ? "Yes" : "No",
             "Peer", btrole.role & ROLE_CONTACT_PEER ? string : "Any" );
    }
    else
      printf("%-12s: *READ ERROR*\n%-12s: *READ ERROR*\n%-12s: *READ ERROR*\n",
             "Role", "Discoverable", "Peer");

    if(bt_read_uart(bt0240, &btuart))
    {
      printf("%-12s: %d\n%-12s: %c\n%-12s: %d\n",
             "BPS Rate", bt_rate2int[(int) btuart.rate],
             "Parity", bt_parity2char[(int) btuart.parity],
             "Stopbits", bt_stopbits2int[(int) btuart.stopbits]);
    }
    else
      printf("%-12s: *READ ERROR*\n%-12s: *READ ERROR*\n%-12s: *READ ERROR*\n",
             "BPS Rate",
             "Parity",
             "Stopbits");
  }

  bt_close(bt0240);

  #if 0
  if(bt0240->name[0] == '/')
  {
    dev_unlock(bt0240->name, getpid());
  }
  #endif
/*  sleep(1); */

  return 0;
}
